### ISSUE TITLE
Perl segfault in regex engine in integer overflow situation

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14858]
Migrated from [rt.perl.org#125826](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=125826) (status was 'resolved')

Searchable as RT125826$

**Comment by p5pRT** [comment]
### From @dcollinsn

Greetings\,

The afl fuzzer has identified the following testcase which causes a segmentation fault in the regular expression engine in perl and miniperl \(oddly\, only when run WITHOUT \-Ilib\)\. The testcase is the 17\-character file&#8203;:

/\\x\{E000000000\}|/

GDB output identifies the segfault within malloc\, which isn't very helpful\, but \(hopefully\) successfuly isolated the crash to Perl\_regexec\_flags\. Valgrind appears to be describing a buffer overflow\. Old versions of perl throw "integer overflow in hexadecimal number" but do not segfault\. Git bisect identifies a commit which appears to be a significant overhaul of part of the regular expression engine\.

\*\*GDB\*\*

GNU gdb \(GDB\) 7\.0\.1\-debian
Copyright \(C\) 2009 Free Software Foundation\, Inc\.
License GPLv3\+&#8203;: GNU GPL version 3 or later \<http&#8203;://gnu\.org/licenses/gpl\.html>
This is free software&#8203;: you are free to change and redistribute it\.
There is NO WARRANTY\, to the extent permitted by law\.  Type "show copying"
and "show warranty" for details\.
This GDB was configured as "i486\-linux\-gnu"\.
For bug reporting instructions\, please see&#8203;:
\<http&#8203;://www\.gnu\.org/software/gdb/bugs/>\.\.\.
Reading symbols from /usr/local/perl\-afl/bin/perl\.\.\.done\.
\(gdb\) run
Starting program&#8203;: /usr/local/perl\-afl/bin/perl allcrash/f4i000012
\[Thread debugging using libthread\_db enabled\]

Program received signal SIGSEGV\, Segmentation fault\.
0xb7e6321e in malloc\_consolidate \(av=\<value optimized out>\) at malloc\.c&#8203;:5153
5153    malloc\.c&#8203;: No such file or directory\.
&nbsp;    in malloc\.c
\(gdb\) bt
\#0  0xb7e6321e in malloc\_consolidate \(av=\<value optimized out>\)
&nbsp;    at malloc\.c&#8203;:5153
\#1  0xb7e654dd in \_int\_malloc \(av=\<value optimized out>\, 
&nbsp;    bytes=\<value optimized out>\) at malloc\.c&#8203;:4373
\#2  0xb7e6797c in \*\_\_GI\_\_\_libc\_malloc \(bytes=4056\) at malloc\.c&#8203;:3661
\#3  0x082c4248 in Perl\_safesysmalloc \(size=4056\) at util\.c&#8203;:149
\#4  0x085cce42 in Perl\_regexec\_flags \(rx=0x873c3d8\, stringarg=0x86df082 ""\, 
&nbsp;    strend=0x86df082 ""\, strbeg=0x86df082 ""\, minend=0\, sv=0x872b268\, 
&nbsp;    data=0x0\, flags=97\) at regexec\.c&#8203;:2979
\#5  0x08366221 in Perl\_pp\_match \(\) at pp\_hot\.c&#8203;:1486
\#6  0x0835d66b in Perl\_runops\_standard \(\) at run\.c&#8203;:41
\#7  0x08106879 in S\_run\_body \(my\_perl=0x8729008\) at perl\.c&#8203;:2448
\#8  perl\_run \(my\_perl=0x8729008\) at perl\.c&#8203;:2371
\#9  0x08065d7e in main \(argc=2\, argv=0xbffff4e4\, env=0xbffff4f0\)
&nbsp;    at perlmain\.c&#8203;:116

\*\*VALGRIND\*\*

==9388== Memcheck\, a memory error detector
==9388== Copyright \(C\) 2002\-2010\, and GNU GPL'd\, by Julian Seward et al\.
==9388== Using Valgrind\-3\.6\.0\.SVN\-Debian and LibVEX; rerun with \-h for copyright info
==9388== Command&#8203;: \.\./bin/perl allcrash/f4i000012
==9388== 
==9388== Invalid write of size 1
==9388==    at 0x85D2ED6&#8203;: Perl\_uvoffuni\_to\_utf8\_flags \(utf8\.c&#8203;:231\)
==9388==    by 0x8213DC9&#8203;: S\_make\_trie \(regcomp\.c&#8203;:2374\)
==9388==    by 0x823F2FD&#8203;: T\.1957 \(regcomp\.c&#8203;:4316\)
==9388==    by 0x828B5CE&#8203;: Perl\_re\_op\_compile \(regcomp\.c&#8203;:7232\)
==9388==    by 0x80D50A7&#8203;: Perl\_pmruntime \(op\.c&#8203;:5579\)
==9388==    by 0x81CE567&#8203;: Perl\_yyparse \(perly\.y&#8203;:1038\)
==9388==    by 0x810F4AE&#8203;: S\_parse\_body \(perl\.c&#8203;:2296\)
==9388==    by 0x81128C8&#8203;: perl\_parse \(perl\.c&#8203;:1626\)
==9388==    by 0x8065B84&#8203;: main \(perlmain\.c&#8203;:114\)
==9388==  Address 0x4237342 is 0 bytes after a block of size 10 alloc'd
==9388==    at 0x4023F50&#8203;: malloc \(vg\_replace\_malloc\.c&#8203;:236\)
==9388==    by 0x82C4247&#8203;: Perl\_safesysmalloc \(util\.c&#8203;:149\)
==9388==    by 0x83C4803&#8203;: Perl\_sv\_grow \(sv\.c&#8203;:1624\)
==9388==    by 0x83C5B72&#8203;: Perl\_newSV \(sv\.c&#8203;:5531\)
==9388==    by 0x8213D97&#8203;: S\_make\_trie \(regcomp\.c&#8203;:2374\)
==9388==    by 0x823F2FD&#8203;: T\.1957 \(regcomp\.c&#8203;:4316\)
==9388==    by 0x828B5CE&#8203;: Perl\_re\_op\_compile \(regcomp\.c&#8203;:7232\)
==9388==    by 0x80D50A7&#8203;: Perl\_pmruntime \(op\.c&#8203;:5579\)
==9388==    by 0x81CE567&#8203;: Perl\_yyparse \(perly\.y&#8203;:1038\)
==9388==    by 0x810F4AE&#8203;: S\_parse\_body \(perl\.c&#8203;:2296\)
==9388==    by 0x81128C8&#8203;: perl\_parse \(perl\.c&#8203;:1626\)
==9388==    by 0x8065B84&#8203;: main \(perlmain\.c&#8203;:114\)
==9388== 
==9388== Invalid write of size 1
==9388==    at 0x85D2EF4&#8203;: Perl\_uvoffuni\_to\_utf8\_flags \(utf8\.c&#8203;:232\)
==9388==    by 0x8213DC9&#8203;: S\_make\_trie \(regcomp\.c&#8203;:2374\)
==9388==    by 0x823F2FD&#8203;: T\.1957 \(regcomp\.c&#8203;:4316\)
==9388==    by 0x828B5CE&#8203;: Perl\_re\_op\_compile \(regcomp\.c&#8203;:7232\)
==9388==    by 0x80D50A7&#8203;: Perl\_pmruntime \(op\.c&#8203;:5579\)
==9388==    by 0x81CE567&#8203;: Perl\_yyparse \(perly\.y&#8203;:1038\)
==9388==    by 0x810F4AE&#8203;: S\_parse\_body \(perl\.c&#8203;:2296\)
==9388==    by 0x81128C8&#8203;: perl\_parse \(perl\.c&#8203;:1626\)
==9388==    by 0x8065B84&#8203;: main \(perlmain\.c&#8203;:114\)
==9388==  Address 0x4237343 is 1 bytes after a block of size 10 alloc'd
==9388==    at 0x4023F50&#8203;: malloc \(vg\_replace\_malloc\.c&#8203;:236\)
==9388==    by 0x82C4247&#8203;: Perl\_safesysmalloc \(util\.c&#8203;:149\)
==9388==    by 0x83C4803&#8203;: Perl\_sv\_grow \(sv\.c&#8203;:1624\)
==9388==    by 0x83C5B72&#8203;: Perl\_newSV \(sv\.c&#8203;:5531\)
==9388==    by 0x8213D97&#8203;: S\_make\_trie \(regcomp\.c&#8203;:2374\)
==9388==    by 0x823F2FD&#8203;: T\.1957 \(regcomp\.c&#8203;:4316\)
==9388==    by 0x828B5CE&#8203;: Perl\_re\_op\_compile \(regcomp\.c&#8203;:7232\)
==9388==    by 0x80D50A7&#8203;: Perl\_pmruntime \(op\.c&#8203;:5579\)
==9388==    by 0x81CE567&#8203;: Perl\_yyparse \(perly\.y&#8203;:1038\)
==9388==    by 0x810F4AE&#8203;: S\_parse\_body \(perl\.c&#8203;:2296\)
==9388==    by 0x81128C8&#8203;: perl\_parse \(perl\.c&#8203;:1626\)
==9388==    by 0x8065B84&#8203;: main \(perlmain\.c&#8203;:114\)
==9388== 
==9388== Invalid write of size 1
==9388==    at 0x85D2F04&#8203;: Perl\_uvoffuni\_to\_utf8\_flags \(utf8\.c&#8203;:233\)
==9388==    by 0x8213DC9&#8203;: S\_make\_trie \(regcomp\.c&#8203;:2374\)
==9388==    by 0x823F2FD&#8203;: T\.1957 \(regcomp\.c&#8203;:4316\)
==9388==    by 0x828B5CE&#8203;: Perl\_re\_op\_compile \(regcomp\.c&#8203;:7232\)
==9388==    by 0x80D50A7&#8203;: Perl\_pmruntime \(op\.c&#8203;:5579\)
==9388==    by 0x81CE567&#8203;: Perl\_yyparse \(perly\.y&#8203;:1038\)
==9388==    by 0x810F4AE&#8203;: S\_parse\_body \(perl\.c&#8203;:2296\)
==9388==    by 0x81128C8&#8203;: perl\_parse \(perl\.c&#8203;:1626\)
==9388==    by 0x8065B84&#8203;: main \(perlmain\.c&#8203;:114\)
==9388==  Address 0x4237344 is 2 bytes after a block of size 10 alloc'd
==9388==    at 0x4023F50&#8203;: malloc \(vg\_replace\_malloc\.c&#8203;:236\)
==9388==    by 0x82C4247&#8203;: Perl\_safesysmalloc \(util\.c&#8203;:149\)
==9388==    by 0x83C4803&#8203;: Perl\_sv\_grow \(sv\.c&#8203;:1624\)
==9388==    by 0x83C5B72&#8203;: Perl\_newSV \(sv\.c&#8203;:5531\)
==9388==    by 0x8213D97&#8203;: S\_make\_trie \(regcomp\.c&#8203;:2374\)
==9388==    by 0x823F2FD&#8203;: T\.1957 \(regcomp\.c&#8203;:4316\)
==9388==    by 0x828B5CE&#8203;: Perl\_re\_op\_compile \(regcomp\.c&#8203;:7232\)
==9388==    by 0x80D50A7&#8203;: Perl\_pmruntime \(op\.c&#8203;:5579\)
==9388==    by 0x81CE567&#8203;: Perl\_yyparse \(perly\.y&#8203;:1038\)
==9388==    by 0x810F4AE&#8203;: S\_parse\_body \(perl\.c&#8203;:2296\)
==9388==    by 0x81128C8&#8203;: perl\_parse \(perl\.c&#8203;:1626\)
==9388==    by 0x8065B84&#8203;: main \(perlmain\.c&#8203;:114\)
==9388== 
==9388== 
==9388== HEAP SUMMARY&#8203;:
==9388==     in use at exit&#8203;: 93\,910 bytes in 591 blocks
==9388==   total heap usage&#8203;: 793 allocs\, 202 frees\, 117\,673 bytes allocated
==9388== 
==9388== LEAK SUMMARY&#8203;:
==9388==    definitely lost&#8203;: 0 bytes in 0 blocks
==9388==    indirectly lost&#8203;: 0 bytes in 0 blocks
==9388==      possibly lost&#8203;: 12\,727 bytes in 292 blocks
==9388==    still reachable&#8203;: 81\,183 bytes in 299 blocks
==9388==         suppressed&#8203;: 0 bytes in 0 blocks
==9388== Rerun with \-\-leak\-check=full to see details of leaked memory
==9388== 
==9388== For counts of detected and suppressed errors\, rerun with&#8203;: \-v
==9388== ERROR SUMMARY&#8203;: 3 errors from 3 contexts \(suppressed&#8203;: 25 from 8\)
dcollins@&#8203;nagios&#8203;:/usr/local/perl\-afl/out$ 

\*\*PERL \-V\*\*

Summary of my perl5 \(revision 5 version 23 subversion 2\) configuration&#8203;:
&nbsp;  Derived from&#8203;: 9728ed0a4dcaca9d7fddf6ce9c5736ed3aacd487
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.6\.32\-5\-686\, archname=i686\-linux\-64int\-ld
&nbsp;    uname='linux nagios 2\.6\.32\-5\-686 \#1 smp tue may 13 16&#8203;:33&#8203;:32 utc 2014 i686 gnulinux '
&nbsp;    config\_args=''
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=undef\, usemultiplicity=undef
&nbsp;    use64bitint=define\, use64bitall=undef\, uselongdouble=define
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='afl\-gcc'\, ccflags ='\-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-g'\,
&nbsp;    cppflags='\-fwrapv \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.4\.5'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=12345678\, doublekind=3
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12\, longdblkind=3
&nbsp;    ivtype='long long'\, ivsize=8\, nvtype='long double'\, nvsize=12\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='afl\-gcc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /usr/lib/gcc/i486\-linux\-gnu/4\.4\.5/include\-fixed /usr/lib /lib/\.\./lib /usr/lib/\.\./lib /lib /usr/lib/i486\-linux\-gnu /usr/lib64
&nbsp;    libs=\-lpthread \-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    perllibs=\-lpthread \-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    libc=libc\-2\.11\.3\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.11\.3'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-g \-L/usr/local/lib \-fstack\-protector'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: HAS\_TIMES PERLIO\_LAYERS PERL\_COPY\_ON\_WRITE
&nbsp;                        PERL\_DONT\_CREATE\_GVSV
&nbsp;                        PERL\_HASH\_FUNC\_ONE\_AT\_A\_TIME\_HARD PERL\_MALLOC\_WRAP
&nbsp;                        PERL\_PRESERVE\_IVUV USE\_64\_BIT\_INT USE\_LARGE\_FILES
&nbsp;                        USE\_LOCALE USE\_LOCALE\_COLLATE USE\_LOCALE\_CTYPE
&nbsp;                        USE\_LOCALE\_NUMERIC USE\_LOCALE\_TIME USE\_LONG\_DOUBLE
&nbsp;                        USE\_PERLIO USE\_PERL\_ATOF
&nbsp;  Locally applied patches&#8203;:
&nbsp;    uncommitted\-changes
&nbsp;  Built under linux
&nbsp;  Compiled at Aug 11 2015 16&#8203;:38&#8203;:21
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /usr/local/perl\-afl/lib/site\_perl/5\.23\.2/i686\-linux\-64int\-ld
&nbsp;    /usr/local/perl\-afl/lib/site\_perl/5\.23\.2
&nbsp;    /usr/local/perl\-afl/lib/5\.23\.2/i686\-linux\-64int\-ld
&nbsp;    /usr/local/perl\-afl/lib/5\.23\.2
&nbsp;    \.

\*\*BISECT\*\*

cdd87c1d4df41f9a54cccff996fa64d291adcee8 is the first bad commit
commit cdd87c1d4df41f9a54cccff996fa64d291adcee8
Author&#8203;: Karl Williamson \<public@&#8203;khwilliamson\.com>
Date&#8203;:   Sun Sep 22 21&#8203;:36&#8203;:29 2013 \-0600

&nbsp;    Teach regex optimizer to handle above\-Latin1
&nbsp;    
&nbsp;    Until this commit\, the regular expression optimizer has essentially
&nbsp;    punted on above\-Latin1 code points\.  Under some circumstances\, they
&nbsp;    would be taken into account\, more or less\, but often\, the generated
&nbsp;    synthetic start class would end up matching all above\-Latin1 code
&nbsp;    points\.  With the advent of inversion lists\, it becomes feasible to
&nbsp;    actually fully handle such code points\, as inversion lists are a
&nbsp;    convenient way to express arbitrary lists of code points and take their
&nbsp;    union\, intersection\, etc\.  This commit changes the optimizer to use
&nbsp;    inversion lists for operating on the code points the synthetic start
&nbsp;    class can match\.
&nbsp;    
&nbsp;    I don't much understand the overall operation of the optimizer\.  I'm
&nbsp;    told that previous porters found that perturbing it caused unexpected
&nbsp;    behaviors\.  I had promised to get this change in 5\.18\, but didn't\.  I'm
&nbsp;    trying to get it in early enough into the 5\.20 preliminary series that
&nbsp;    any problems will surface before 5\.20 ships\.
&nbsp;    
&nbsp;    This commit doesn't change the macro level logic\, but does significantly
&nbsp;    change various micro level things\.  Thus the 'and' and 'or' subroutines
&nbsp;    have been rewritten to use inversion lists\.  I'm pretty confident that
&nbsp;    they do what their names suggest\.  I re\-derived the equations for what
&nbsp;    these operations should do\, getting the same results in some cases\, but
&nbsp;    extending others where the previous code mostly punted\.  The derivations
&nbsp;    are given in comments in the respective routines\.
&nbsp;    
&nbsp;    Some of the code is greatly simplified\, as it no longer has to treat
&nbsp;    above\-Latin1 specially\.
&nbsp;    
&nbsp;    It is now feasible for /i matching of above\-Latin1 code points to know
&nbsp;    explicitly the folds that should be in the synthetic start class\.  But
&nbsp;    more prepatory work needs to be done before putting that into place\.
&nbsp;    \.\.\.

:100644 100644 ec203f9c1f3ea42c65324e632c746042c32954f1 3dd62f946eedd99f87489fecfeb1acd86e2d250b M    embed\.fnc
:100644 100644 fca8736feb457d00987232649f8b563aef4f24fa 5fc9171d233fd06b4d6cc08570cdb41f976d0b0b M    embed\.h
:100644 100644 568cdf733c7784cc3f5a059583d9ae9a262ae72a 33d00ef42cf01836e9365e4a9086cf64b678b74b M    proto\.h
:100644 100644 ec24583f1f7bec42e186a4241353b9688bf6f2de efefd0a17ba14144286dc55c149d6ee422b18249 M    regcomp\.c
:100644 100644 f0153fc12c842c26d928d3303feb519430b73a07 eccb46690a4d252fb107044116973e3baa0cbd2c M    regcomp\.h
bisect run success

**Comment by p5pRT** [comment]
### From @tonycoz

On Sun Aug 16 20&#8203;:56&#8203;:17 2015\, dcollinsn@&#8203;gmail\.com wrote&#8203;:
> Greetings\,
> 
> The afl fuzzer has identified the following testcase which causes a
> segmentation fault in the regular expression engine in perl and
> miniperl \(oddly\, only when run WITHOUT \-Ilib\)\. The testcase is the 17\-
> character file&#8203;:
> 
> /\\x\{E000000000\}|/

Here's a fix \(attached\)\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

<details><summary>0001-perl-125826-make-the-buffer-large-enough-in-TRIE_STO.patch</summary>

```diff
From ab3f825e8c3d0b6f70faac9d6b3552923bd511d0 Mon Sep 17 00:00:00 2001
From: Tony Cook <tony@develop-help.com>
Date: Tue, 18 Aug 2015 12:11:12 +1000
Subject: [PATCH] [perl #125826] make the buffer large enough in
 TRIE_STORE_REVCHAR

---
 regcomp.c           | 2 +-
 t/re/pat_advanced.t | 9 +++++++++
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/regcomp.c b/regcomp.c
index f08f08f..c052cc7 100644
--- a/regcomp.c
+++ b/regcomp.c
@@ -2001,7 +2001,7 @@ is the recommended Unicode-aware way of saying
 #define TRIE_STORE_REVCHAR(val)                                            \
     STMT_START {                                                           \
 	if (UTF) {							   \
-            SV *zlopp = newSV(7); /* XXX: optimize me */                   \
+            SV *zlopp = newSV(UTF8_MAXBYTES); /* XXX: optimize me */       \
 	    unsigned char *flrbbbbb = (unsigned char *) SvPVX(zlopp);	   \
             unsigned const char *const kapow = uvchr_to_utf8(flrbbbbb, val); \
 	    SvCUR_set(zlopp, kapow - flrbbbbb);				   \
diff --git a/t/re/pat_advanced.t b/t/re/pat_advanced.t
index 230fd89..33647f3 100644
--- a/t/re/pat_advanced.t
+++ b/t/re/pat_advanced.t
@@ -2419,6 +2419,15 @@ EOF
                         'No segfault on qr{(?&foo){0}abc(?<foo>)}');
     }
 
+    SKIP:
+    {   # [perl #125826] buffer overflow in TRIE_STORE_REVCHAR
+        # (during compilation, so use a fresh perl)
+        $Config{uvsize} == 8
+	  or skip("need large code-points for this test", 1);
+	fresh_perl_is('/\x{E000000000}|/ and print qq(ok\n)', "ok\n", {},
+		      "buffer overflow in TRIE_STORE_REVCHAR");
+    }
+
     # !!! NOTE that tests that aren't at all likely to crash perl should go
     # a ways above, above these last ones.
 
-- 
2.5.0


```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @dcollinsn

Hello p5p\,

Just wondering if these are useful enough that you'd like me to continue
reporting them\. I've had a passing interest in the perl ecosystem for a few
years now\, and have been hoping to find enough time to learn the codebase
so that I can actually contribute\, but in the mean time I was experimenting
with afl for fun\. Is there any information I should provide to make these
reports more useful\, and do they hold any value at all beyond corner cases
unlikely to ever be encountered in the real world?

Regards\,
Dan Collins

On Mon\, Aug 17\, 2015 at 10&#8203;:18 PM\, Tony Cook via RT \<
perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> On Sun Aug 16 20&#8203;:56&#8203;:17 2015\, dcollinsn@&#8203;gmail\.com wrote&#8203;:
> > Greetings\,
> >
> > The afl fuzzer has identified the following testcase which causes a
> > segmentation fault in the regular expression engine in perl and
> > miniperl \(oddly\, only when run WITHOUT \-Ilib\)\. The testcase is the 17\-
> > character file&#8203;:
> >
> > /\\x\{E000000000\}|/
>
> Here's a fix \(attached\)\.
>
> Tony
>
> \-\-\-
> via perlbug&#8203;:  queue&#8203;: perl5 status&#8203;: new
> https://rt-archive.perl.org/perl5/Ticket/Display.html?id=125826
>
> From ab3f825e8c3d0b6f70faac9d6b3552923bd511d0 Mon Sep 17 00&#8203;:00&#8203;:00 2001
> From&#8203;: Tony Cook \<tony@&#8203;develop\-help\.com>
> Date&#8203;: Tue\, 18 Aug 2015 12&#8203;:11&#8203;:12 \+1000
> Subject&#8203;: \[PATCH\] \[perl \#125826\] make the buffer large enough in
>  TRIE\_STORE\_REVCHAR
>
> \-\-\-
>  regcomp\.c           | 2 \+\-
>  t/re/pat\_advanced\.t | 9 \+\+\+\+\+\+\+\+\+
>  2 files changed\, 10 insertions\(\+\)\, 1 deletion\(\-\)
>
> diff \-\-git a/regcomp\.c b/regcomp\.c
> index f08f08f\.\.c052cc7 100644
> \-\-\- a/regcomp\.c
> \+\+\+ b/regcomp\.c
> @&#8203;@&#8203; \-2001\,7 \+2001\,7 @&#8203;@&#8203; is the recommended Unicode\-aware way of saying
>  \#define TRIE\_STORE\_REVCHAR\(val\)
>   \\
>      STMT\_START \{
>  \\
>         if \(UTF\) \{
>  \\
> \-            SV \*zlopp = newSV\(7\); /\* XXX&#8203;: optimize me \*/
>  \\
> \+            SV \*zlopp = newSV\(UTF8\_MAXBYTES\); /\* XXX&#8203;: optimize me \*/
>  \\
>             unsigned char \*flrbbbbb = \(unsigned char \*\) SvPVX\(zlopp\);
> \\
>              unsigned const char \*const kapow = uvchr\_to\_utf8\(flrbbbbb\,
> val\); \\
>             SvCUR\_set\(zlopp\, kapow \- flrbbbbb\);
> \\
> diff \-\-git a/t/re/pat\_advanced\.t b/t/re/pat\_advanced\.t
> index 230fd89\.\.33647f3 100644
> \-\-\- a/t/re/pat\_advanced\.t
> \+\+\+ b/t/re/pat\_advanced\.t
> @&#8203;@&#8203; \-2419\,6 \+2419\,15 @&#8203;@&#8203; EOF
>                          'No segfault on qr\{\(?&foo\)\{0\}abc\(?\<foo>\)\}'\);
>      \}
>
> \+    SKIP&#8203;:
> \+    \{   \# \[perl \#125826\] buffer overflow in TRIE\_STORE\_REVCHAR
> \+        \# \(during compilation\, so use a fresh perl\)
> \+        $Config\{uvsize\} == 8
> \+         or skip\("need large code\-points for this test"\, 1\);
> \+       fresh\_perl\_is\('/\\x\{E000000000\}|/ and print qq\(ok\\n\)'\, "ok\\n"\, \{\}\,
> \+                     "buffer overflow in TRIE\_STORE\_REVCHAR"\);
> \+    \}
> \+
>      \# \!\!\! NOTE that tests that aren't at all likely to crash perl should
> go
>      \# a ways above\, above these last ones\.
>
> \-\-
> 2\.5\.0
>
>
>

**Comment by p5pRT** [comment]
### From PeterCMartini@GMail.com

On Mon\, Aug 17\, 2015 at 11&#8203;:28 PM\, Dan Collins \<dcollinsn@&#8203;gmail\.com> wrote&#8203;:

> Hello p5p\,
>
> Just wondering if these are useful enough that you'd like me to continue
> reporting them\. I've had a passing interest in the perl ecosystem for a few
> years now\, and have been hoping to find enough time to learn the codebase
> so that I can actually contribute\, but in the mean time I was experimenting
> with afl for fun\. Is there any information I should provide to make these
> reports more useful\, and do they hold any value at all beyond corner cases
> unlikely to ever be encountered in the real world?
>

I can only speak for myself\, but yes\, I find these very useful; those
corner cases are the edges we can investigate first during refactoring\, and
also the places to be most careful about if any refactoring is being done\.

The cases I've seen you provide have been solid enough that I wouldn't ask
for any more details\, though I'd also be very happy to get some more
details on how you've set up AFL to find these :\-\)


>
> Regards\,
> Dan Collins
>
> On Mon\, Aug 17\, 2015 at 10&#8203;:18 PM\, Tony Cook via RT \<
> perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>
>> On Sun Aug 16 20&#8203;:56&#8203;:17 2015\, dcollinsn@&#8203;gmail\.com wrote&#8203;:
>> > Greetings\,
>> >
>> > The afl fuzzer has identified the following testcase which causes a
>> > segmentation fault in the regular expression engine in perl and
>> > miniperl \(oddly\, only when run WITHOUT \-Ilib\)\. The testcase is the 17\-
>> > character file&#8203;:
>> >
>> > /\\x\{E000000000\}|/
>>
>> Here's a fix \(attached\)\.
>>
>> Tony
>>
>> \-\-\-
>> via perlbug&#8203;:  queue&#8203;: perl5 status&#8203;: new
>> https://rt-archive.perl.org/perl5/Ticket/Display.html?id=125826
>>
>> From ab3f825e8c3d0b6f70faac9d6b3552923bd511d0 Mon Sep 17 00&#8203;:00&#8203;:00 2001
>> From&#8203;: Tony Cook \<tony@&#8203;develop\-help\.com>
>> Date&#8203;: Tue\, 18 Aug 2015 12&#8203;:11&#8203;:12 \+1000
>> Subject&#8203;: \[PATCH\] \[perl \#125826\] make the buffer large enough in
>>  TRIE\_STORE\_REVCHAR
>>
>> \-\-\-
>>  regcomp\.c           | 2 \+\-
>>  t/re/pat\_advanced\.t | 9 \+\+\+\+\+\+\+\+\+
>>  2 files changed\, 10 insertions\(\+\)\, 1 deletion\(\-\)
>>
>> diff \-\-git a/regcomp\.c b/regcomp\.c
>> index f08f08f\.\.c052cc7 100644
>> \-\-\- a/regcomp\.c
>> \+\+\+ b/regcomp\.c
>> @&#8203;@&#8203; \-2001\,7 \+2001\,7 @&#8203;@&#8203; is the recommended Unicode\-aware way of saying
>>  \#define TRIE\_STORE\_REVCHAR\(val\)
>>   \\
>>      STMT\_START \{
>>    \\
>>         if \(UTF\) \{
>>  \\
>> \-            SV \*zlopp = newSV\(7\); /\* XXX&#8203;: optimize me \*/
>>    \\
>> \+            SV \*zlopp = newSV\(UTF8\_MAXBYTES\); /\* XXX&#8203;: optimize me \*/
>>    \\
>>             unsigned char \*flrbbbbb = \(unsigned char \*\) SvPVX\(zlopp\);
>>   \\
>>              unsigned const char \*const kapow = uvchr\_to\_utf8\(flrbbbbb\,
>> val\); \\
>>             SvCUR\_set\(zlopp\, kapow \- flrbbbbb\);
>>   \\
>> diff \-\-git a/t/re/pat\_advanced\.t b/t/re/pat\_advanced\.t
>> index 230fd89\.\.33647f3 100644
>> \-\-\- a/t/re/pat\_advanced\.t
>> \+\+\+ b/t/re/pat\_advanced\.t
>> @&#8203;@&#8203; \-2419\,6 \+2419\,15 @&#8203;@&#8203; EOF
>>                          'No segfault on qr\{\(?&foo\)\{0\}abc\(?\<foo>\)\}'\);
>>      \}
>>
>> \+    SKIP&#8203;:
>> \+    \{   \# \[perl \#125826\] buffer overflow in TRIE\_STORE\_REVCHAR
>> \+        \# \(during compilation\, so use a fresh perl\)
>> \+        $Config\{uvsize\} == 8
>> \+         or skip\("need large code\-points for this test"\, 1\);
>> \+       fresh\_perl\_is\('/\\x\{E000000000\}|/ and print qq\(ok\\n\)'\, "ok\\n"\, \{\}\,
>> \+                     "buffer overflow in TRIE\_STORE\_REVCHAR"\);
>> \+    \}
>> \+
>>      \# \!\!\! NOTE that tests that aren't at all likely to crash perl should
>> go
>>      \# a ways above\, above these last ones\.
>>
>> \-\-
>> 2\.5\.0
>>
>>
>>
>

**Comment by p5pRT** [comment]
### From @tonycoz

On Mon\, Aug 17\, 2015 at 11&#8203;:28&#8203;:03PM \-0400\, Dan Collins wrote&#8203;:
> Hello p5p\,
> 
> Just wondering if these are useful enough that you'd like me to continue
> reporting them\. I've had a passing interest in the perl ecosystem for a few
> years now\, and have been hoping to find enough time to learn the codebase
> so that I can actually contribute\, but in the mean time I was experimenting
> with afl for fun\. Is there any information I should provide to make these
> reports more useful\, and do they hold any value at all beyond corner cases
> unlikely to ever be encountered in the real world?

I think they're useful\, even if we don't always have the development
tuits to deal with them immediately\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

On Mon Aug 17 19&#8203;:18&#8203;:42 2015\, tonyc wrote&#8203;:
> On Sun Aug 16 20&#8203;:56&#8203;:17 2015\, dcollinsn@&#8203;gmail\.com wrote&#8203;:
> > Greetings\,
> > 
> > The afl fuzzer has identified the following testcase which causes a
> > segmentation fault in the regular expression engine in perl and
> > miniperl \(oddly\, only when run WITHOUT \-Ilib\)\. The testcase is the 17\-
> > character file&#8203;:
> > 
> > /\\x\{E000000000\}|/
> 
> Here's a fix \(attached\)\.

Applied with non\-code changes as 668fcfea69086ab3cf35fc7ba175ea225dbc7f9d\.

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @rurban


> On Aug 18\, 2015\, at 5&#8203;:29 AM\, Dan Collins via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> 
> Hello p5p\,
> 
> Just wondering if these are useful enough that you'd like me to continue
> reporting them\. I've had a passing interest in the perl ecosystem for a few
> years now\, and have been hoping to find enough time to learn the codebase
> so that I can actually contribute\, but in the mean time I was experimenting
> with afl for fun\. Is there any information I should provide to make these
> reports more useful\, and do they hold any value at all beyond corner cases
> unlikely to ever be encountered in the real world?

I certainly do not speak for p5p\, they usually tend to go into the opposite direction whenever I voice my opinion\,
but I do value your afl reports a lot\, and I try to fix all of them\. I track them in a special ticket and most of them 
are already fixed\. They are not getting lost\.
Even if some p5p committers reject them as unfixable or such\, I fix them at least in our version of perl5 for the 
benefit of some other perl5 users\.

They are useful as is\. No need to worry\.
Thanks\.

My list \(maybe I missed some lately\)&#8203;:

New crashes\, not yet in the 5\.20\.3 blockers list&#8203;:
&nbsp;    • perl \#125341\, duplicate of perl \#121048 \(Fixed only with bugfix/CM\-819\-rt125341\-begin\)
&nbsp;    • perl \#125697 fixed in 5\.22
&nbsp;    • perl \#125540 S\_scan\_heredoc&#8203;: Assertion \`s' failed at toke\.c&#8203;:9314 \(Todo\)
&nbsp;    • perl \#125534 Perl\_sv\_clear&#8203;: Assertion \(SvTYPE\(sv\) \!= \(svtype\)SVTYPEMASK\) failed \(sv\.c&#8203;:6395\) with \-e'map\{%0=map\{0\}m 0 0\}%0=map\{0\}0' \(Todo\)
&nbsp;    • perl \#125350 null ptr deref \-> S\_clear\_yystack \(perly\.c&#8203;:218\) \(Todo\)
&nbsp;    • perl \#123878 Infinite recursion \(\+segfault\) on die\(\) after goto\-ing \(Todo\, very relevant to us\!\) See \`bugfix/CM\-891\-rt123878\-goto\-die\`
&nbsp;    • perl \#125840 \-e'$x=\*0; \*x=$x' glob\_assign\_glob inner ref \(Rejected by p5p\. Fixed in merge\-upstream\, with bugfix/CM\-891\-glob\_assign\_glob\-crash\)


see http&#8203;://perl5\.git\.perl\.org/perl\.git/blob/refs/heads/maint\-5\.20\-votes&#8203;:/Porting/cherry\-pick\-votes\-maint\-5\.20\.xml
or Tickets Listed in \#123921&#8203;: 5\.20\.3 blockers

&nbsp;    • perl \#123554 avoid a crash from SvGROW\(MEM\_SIZE\_MAX\) \(Fixed in 5\.22\)
&nbsp;    • Fix double free with const overload after errors \(Fixed in 5\.22 with 67c71cbbd62a7\)
&nbsp;    • perl \#123617 Localise PL\_lex\_stuff \(crash fix\) \(Fixed in 5\.22 with eabab8bccf8\)
&nbsp;    • perl \#123955 Fix assert fail with 0 s/// in quotes \(Fixed in 5\.22 with ce7c414ee71\)
&nbsp;    • perl \#123737 Fix assertion failure with 0$\{ \(Fixed in 5\.22 with 488bc57958911\)
&nbsp;    • perl \#123737 Fix assertion failure with 0$\#\{ \(Fixed in 5\.22 with 310a0d0a370ea\)
&nbsp;    • perl \#123753 &\\0foo parsing \(Fixed in 5\.22 with 3c47da3c2e\)
&nbsp;    • perl \#123753 Assert fail with &\{\+foo\} and errors \(Fixed in 5\.22 with eea89386b3f\)
&nbsp;    • perl \#123677 Crash with token stack overflow \(Fixed in 5\.22 with 7aa8cb0dec173d\)
&nbsp;    • perl \#123759 always count on OPpTRANS\_IDENTICAL \(Fixed in 5\.22 with a53bfdae91fb2d7\)
&nbsp;    • perl \#123755 including unknown char in error requires care \(Fixed in 5\.22 with 8a6d8ec6fe62\)
regcomp can read past end of string after parsing flags
&nbsp;    • perl \#123554 catch a couple of other size overflows \(Fixed in 5\.22 with 4 commits\)
&nbsp;    • perl \#123554 fix threaded builds and prevent a warning \(Fixed in 5\.22\)
&nbsp;    • perl \#123816 fix stat stacking \(Fixed in 5\.22 with 87ebf1e3df\)
&nbsp;    • perl \#123870 fixup trie runtime debug output \(Fixed in 5\.22 with d0bec203\)
&nbsp;    • perl \#123874 fix argument underflow for pack\(\) \(Fixed in 5\.22 with fc1bb3f2dcaa\)
&nbsp;    • perl \#123801 Stop s\#\#\[\}\#e from crashing \(Fixed in 5\.22 with f4460c6f7a0de1\)
&nbsp;    • perl \#123802 Assertion failure with /$0\{\}/ \(Fixed in 5\.22 with 479ae48e22f11\)
&nbsp;    • perl \#123802 Assertion failure with qq\[\\L\\L\] \(Fixed in 5\.22 with 66edcf79f\)
&nbsp;    • perl \#123763 Clear target on my $\_; split \(Fixed in 5\.22 with 55b3980349\)
&nbsp;    • perl \#123817 Assert fail with attr in anon hash \(Fixed in 5\.22 with 6b2b48acae2d7\)
&nbsp;    • perl \#123849 sv\.c&#8203;: Fix sv\_clear \-Do output \(Fixed in 5\.22 with 813d2eb1716\)
&nbsp;    • perl \#123960 sv\.c&#8203;: Fix gp\_free \-Do output \(Fixed in 5\.22 with 923ed5809\)
&nbsp;    • perl \#123963 qq\[@&#8203;\<fullwidth digit>\] \(Fixed in 5\.22 with 9d58dbc453a\)
&nbsp;    • perl \#123847 crash with \*foo&#8203;::=\*bar&#8203;::=\*with\_hash \(Fixed in 5\.22 with 3d50185de85a\)
&nbsp;    • perl \#123995 Assert fail with s;@&#8203;\{\<\<; \(Fixed in 5\.22 with b24768f02e4a\)
&nbsp;    • perl \#123790 Assert fail with \*x=\<y> \(Fixed in 5\.22 with aab1202a8d4b6 and 21639bf469a0\)
&nbsp;    • perl \#124099 Wrong CvOUTSIDE in find\_lexical\_cv \(Fixed in 5\.22 with d655d9a2c4d4884\)
&nbsp;    • perl \#124187 don't call pad\_findlex\(\) on a NULL CV \(Fixed in 5\.22 with b12396ac84\)
&nbsp;    • perl \#124385 null ptr deref in Perl\_cv\_forget\_slab \(Fixed in 5\.22 with de0885da32\)

more crashes&#8203;:
&nbsp;    • perl \#123711 Fix crash with 0\-5x\-l \{0\} \(Fixed with 5\.22 with 83a85f49e26\)
&nbsp;    • perl \#123712 Fix /$a\[/ parsing \(Fixed with 5\.22 with e47d32dcd59a5782\)
&nbsp;    • perl \#123712 Don't check sub\_inwhat \(Fixed with 5\.22 with d27f4b916ce\)
&nbsp;    • perl \#124156 death during unwinding causes crash \(Fixed with 5\.22 with 1956db7ee60\)
&nbsp;    • perl \#123398\) don't fatalize warnings during unwinding \(Fixed in our perl with 46b27d2f2c3\, reportedly merged into 5\.22 lately\)
